### PR TITLE
cdc: check observe id before starting to observe

### PR DESCRIPTION
**Pull request by 5kbpers**
### What problem does this PR solve?

Problem Summary:
Sometimes the ordering of read requests will be changed.

### What is changed and how it works?
Check observe id for rejecting stale observe cmd.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 
N/A

### CODE REVIEW

File: components/raftstore/src/store/fsm/apply.rs [hunk 1]
```diff
@@ -2444,7 +2444,7 @@ impl Debug for GenSnapTask {
 static OBSERVE_ID_ALLOC: AtomicUsize = AtomicUsize::new(0);
 
 /// A unique identifier for checking stale observed commands.
-#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
```
Comment (overvenus) [inline components/raftstore/src/store/fsm/apply.rs:2447]: We need to use `Ordering::SeqCst` to new `ObserveID` if we use the value to determine staleness between different `ObserveID`s.

Comment (5kbpers) [reply components/raftstore/src/store/fsm/apply.rs:2447]: I don't think so, all the updates and comparisons would happen in a single thread.

Comment (overvenus) [reply components/raftstore/src/store/fsm/apply.rs:2447]: But still, it is a pitfall for futures improvement (eg, multi-thread CDC).

File: components/cdc/tests/failpoints/test_observe.rs [hunk 2]
```diff
@@ -125,3 +133,85 @@ fn test_stale_observe_cmd() {
     event_feed_wrap.as_ref().replace(None);
     suite.stop();
 }
+
+#[test]
+fn test_delayed_change_cmd() {
```
Comment (overvenus) [inline components/cdc/tests/failpoints/test_observe.rs:138]: What does it want to test?

File: components/raftstore/src/store/fsm/apply.rs [hunk 3]
```diff
@@ -2901,25 +2901,18 @@ where
...
-            }
+            } => (observe_id, region_id, Some(enabled)),
             ChangeCmd::Snapshot {
                 observe_id,
                 region_id,
             } => (observe_id, region_id, None),
         };
+        if let Some(observe_cmd) = self.delegate.observe_cmd.as_mut() {
+            if observe_cmd.id > observe_id {
+                notify_stale_req(self.delegate.term, cb);
```
Comment (overvenus) [inline components/raftstore/src/store/fsm/apply.rs:2912]: Stale error will be translated into region not found error is it ok?
https://github.com/tikv/tikv/blob/cdca278ba7c2f631d5b3b5db08f1335dc7a89108/components/cdc/src/delegate.rs#L292-L311

File: components/cdc/tests/failpoints/test_observe.rs [hunk 4]
```diff
@@ -125,3 +133,85 @@ fn test_stale_observe_cmd() {
...
+    // Wait util lease expired
+    sleep_ms(300);
+
+    let (sx, rx) = mpsc::sync_channel::<RaftMessage>(1);
+    let send_flag = Arc::new(AtomicBool::new(true));
+    let send_read_index_filter = RegionPacketFilter::new(region.get_id(), leader.get_store_id())
+        .direction(Direction::Send)
+        .msg_type(MessageType::MsgHeartbeat)
+        .set_msg_callback(Arc::new(move |msg: &RaftMessage| {
+            if send_flag.compare_and_swap(true, false, Ordering::SeqCst) {
```
Comment (overvenus) [inline components/cdc/tests/failpoints/test_observe.rs:159]: Why send heartbeat once?

File: components/cdc/tests/failpoints/test_observe.rs (hunk 2 above)
Comment (5kbpers) [reply components/cdc/tests/failpoints/test_observe.rs:138]: For testing change cmd which contains a stale observe id.

File: components/raftstore/src/store/fsm/apply.rs (hunk 3 above)
Comment (5kbpers) [reply components/raftstore/src/store/fsm/apply.rs:2912]: It is ok for now, the client will treat those errors in a same way.

File: components/cdc/tests/failpoints/test_observe.rs (hunk 4 above)
Comment (5kbpers) [reply components/cdc/tests/failpoints/test_observe.rs:159]: Because thread will be blocked if there is not any receiver.

**Pull request merged**